src/Compilers/Core/Portable/SourceGeneration/SourceGeneratorContext.cs

```diff
@@ -0,0 +1,60 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Text;
+using System.Threading;
+using Microsoft.CodeAnalysis.Diagnostics;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Text;
+using Roslyn.Utilities;
+
+namespace Microsoft.CodeAnalysis
+{
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0016:Add public types and members to the declared API", Justification = "In Progress")]
+    public readonly struct SourceGeneratorContext
+    {
+        internal SourceGeneratorContext(Compilation compilation, AnalyzerOptions options, CancellationToken cancellationToken = default)
+        {
+            Compilation = compilation;
+            AnalyzerOptions = options;
+            CancellationToken = cancellationToken;
+            AdditionalSources = new AdditionalSourcesCollection();
+        }
+
+        public Compilation Compilation { get; }
+
+        // PROTOTYPE: replace AnalyzerOptions with an differently named type that is otherwise identical.
+        // The concern being that something added to one isn't necessarily applicable to the other.
+        public AnalyzerOptions AnalyzerOptions { get; }
+
+        public CancellationToken CancellationToken { get; }
+
+        public AdditionalSourcesCollection AdditionalSources { get; }
+
+        public void ReportDiagnostic(Diagnostic diagnostic) { throw new NotImplementedException(); }
+    }
+
+    [System.Diagnostics.CodeAnalysis.SuppressMessage("ApiDesign", "RS0016:Add public types and members to the declared API", Justification = "In progress")]
+    // PROTOTYPE: this is going to need to track the input and output compilations that occured
+    public struct UpdateContext
+    {
+        internal UpdateContext(ImmutableArray<GeneratedSourceText> sources, CancellationToken cancellationToken = default)
+        {
+            AdditionalSources = new AdditionalSourcesCollection(sources);
+            Succeeded = true;
+            CancellationToken = cancellationToken;
+        }
+
+        public CancellationToken CancellationToken { get; }
+
+        public AdditionalSourcesCollection AdditionalSources { get; }
+
+        public bool Succeeded { get; set; }
```

Comment:
All of the existing APIs which take an `UpdateContext` end up also returning a new one. Given that it would seem that a `readonly struct` would be more appropriate here. Is there a reason for keeeping this mutable?

Reply:
Reason was the `Succeeded` field. 

It's perfectly valid that a generator might not be able to apply an edit; it's not an error per-se, it just means that it needs to do a full regeneration. Thinking about it, making the user return the updated context is actually kind of weird anyway. I'll seal it and make the generator return a bool instead.

---
In reply to: [381653056](https://github.com/dotnet/roslyn/pull/41789#discussion_r381653056) [](ancestors = 381653056)